### PR TITLE
Update base images reference page

### PR DESCRIPTION
**Pull request by nghiant2710**
Connects-to: #1902 
Dynamically generate content from api v6/device_type endpoint

Change-type: patch
Signed-off-by: Trong Nghia Nguyen <nghiant2710@gmail.com>

---

> *Please make sure to read the [CONTRIBUTING](https://github.com/balena-io/docs/blob/master/CONTRIBUTING.md) document before opening the PR for relevant information on contributing to the documentation. Thanks!*

### CODE REVIEW

File: templates/_base-images-ref.html
Comment (thgreasi) [inline templates/_base-images-ref.html:7]:
```suggestion
        url: "https://api.balena-cloud.com/v6/device_type?$select=name,slug&$expand=is_of__cpu_architecture($select=slug)&$filter=is_default_for__application/any(idfa:((idfa/is_host%20eq%20true)%20and%20(idfa/is_archived%20eq%20false))%20and%20not(contains(idfa/app_name,%27esr%27)))&$orderby=name%20asc"
```

File: templates/_base-images-ref.html
Comment (thgreasi) [inline templates/_base-images-ref.html:36]: Let's drop this after adding an $orderby in the OData query above.

**Comment by nghiant2710** [comment]
@thgreasi @vipulgupta2048 I updated all your comments, can you please check them again?

File: templates/_base-images-ref.html
Comment (thgreasi) [inline templates/_base-images-ref.html:11]: Heads-up that only the following are in the list of returned items: 'n310-tx2', 'etcher-pro'.
Any chance the rest are just aliases? eg `nuc` is an alias and the actual DT slug is `intel-nuc`
Here is the actual slug list:
```
["n310-tx2", "apalis-imx6q", "colibri-imx6dl", "hummingboard", "edge", "etcher-pro", "intel-edison", "hummingboard", "odroid-xu4", "odroid-xu4", "intel-nuc", "parallella", "hummingboard"]
```
which I got by doing
```
await Promise.all(['n310-tx2', 'apalis-imx6', 'colibri-imx6', 'cubox-i', 'edge', 'etcher-pro', 'edison', 'hummingboard2', 'odroid-u3+', 'odroid-ux3', 'nuc', 'parallella-hdmi-resin', 'solidrun-imx6'].map(dt => sdk.models.device.getDeviceSlug(dt)))
```

Now the extra question on top of this is why do we actually need to exclude those?
Probably better to add as a comment so that we don't forget.

File: templates/_base-images-ref.html
Comment (thgreasi) [inline templates/_base-images-ref.html:13]:
```suggestion
        var deviceTypesByArch = {};
```
right?

File: templates/_base-images-ref.html
Comment (thgreasi) [inline templates/_base-images-ref.html:11]: Heads-up that only the following are in the list of returned items: 'n310-tx2', 'etcher-pro'.
Any chance the rest are just aliases? eg `nuc` is an alias and the actual DT slug is `intel-nuc`
Here is the actual slug list:
```
["n310-tx2", "apalis-imx6q", "colibri-imx6dl", "hummingboard", "edge", "etcher-pro", "intel-edison", "hummingboard", "odroid-xu4", "odroid-xu4", "intel-nuc", "parallella", "hummingboard"]
```
which I got by doing
```
await Promise.all(['n310-tx2', 'apalis-imx6', 'colibri-imx6', 'cubox-i', 'edge', 'etcher-pro', 'edison', 'hummingboard2', 'odroid-u3+', 'odroid-ux3', 'nuc', 'parallella-hdmi-resin', 'solidrun-imx6'].map(dt => sdk.models.device.getDeviceSlug(dt)))
```

Now the extra question on top of this is why do we actually need to exclude those?
Probably better to add as a comment so that we don't forget.

File: templates/_base-images-ref.html
Comment (thgreasi) [inline templates/_base-images-ref.html:27]:
```suggestion
        })
```

File: templates/_base-images-ref.html
Comment (thgreasi) [inline templates/_base-images-ref.html:83]: How about changing `generateXYZTable` to use a generic `generateTable(deviceType, tableDefinitionOptions /* keyof typeof tableDefinitions */)` and abstract their implementation into the following definitions:
```js
var repoUrlsByLanguage = {
	python: $links.githubPythonBaseImages,
	golang: $links.githubGolangBaseImages,
	openjdk: $links.githubOpenJDKBaseImages,
	node: $links.githubNodeBaseImages,
	dontnet: $links.githubDotnetBaseImages,
};
var languages = [
	'python',
	'golang',
	'openjdk',
	'node',
];
var tableDefinitions = {
	alpine: {
		baseImagePrefixes: ['alpine']
		versions: '3.13 (latest), 3.11, 3.10, 3.12, edge',
		languages: languages,
	},
	armv6: {
		baseImagePrefixes : ['']
		versions: 'buster (latest), bullseye, jessie, stretch'
	},
	debian: {
		baseImagePrefixes : ['', 'debian']
		versions: 'buster (latest), bullseye, jessie, stretch'
		languages: languages.concat('dontnet'),
	},
	ubuntu: {
		baseImagePrefixes : ['ubuntu']
		versions: 'bionic (latest), cosmic, xenial, disco, eoan, focal'
		languages: languages,
	}
};
```
We can do that at a later point of course, but in that case lets open an issue referencing this comment.

File: templates/_base-images-ref.html
Comment (nghiant2710) [reply templates/_base-images-ref.html:11]: @thgreasi, I create that list based on this file https://docs.google.com/spreadsheets/d/1AX0r5P-JuL4tllFUJNEURYpgazrWQ6aq2Ac94Yev5Mg/edit#gid=1120827262, basically all devices which are not in our contracts, no longer supported will be in the list. I'm not sure what will be returned by the `device-type` endpoint so will list all the devices without base images in the list.

Comment (thgreasi) [reply templates/_base-images-ref.html:11]: @nghiant2710 it feels weird that we have device types that we support for which we don't have base images for. Anything holding us back from releases base images for those as well?
Discontinued DTs should not be returned by the above OData query, since we should be marking them as archived in the DB. 
I went through them and I think the only ones that are still in the list are the following:
```suggestion
        // Device Types for which we don't have base images
        var IGNORE_THESE_DEVICES = ["n310-tx2", "etcher-pro", "intel-edison", "odroid-xu4", "intel-nuc"];
```

Comment (nghiant2710) [reply templates/_base-images-ref.html]: @thgreasi we don't create the base images for them because they're not in our contracts. I updated the list based on your suggestion (`intel-edison`, `intel-nuc` removed since they're in contracts)

File: templates/_base-images-ref.html
Comment (nghiant2710) [reply templates/_base-images-ref.html]: Renamed the variable.

I think just using `forEach` loop with the DTs list is better here since we can add some extra logic to exclude the device in ignore list or replace some contents in the callback. `_.groupBy` is fine but we need to add extra steps after grouping them to the above tasks.

File: templates/_base-images-ref.html
Comment (nghiant2710) [reply templates/_base-images-ref.html:84]: There are 2 kinds of generateXYZTable functions here, while `generateDebian|Ubuntu|Fedora|AlpineTable` are generic functions that generate contents for each distributions, the others will call the generic ones to correctly generate contents for it.

I think your suggestion will really help improve the current generic functions as we still have few things can be abstracted but it's better to move this to another PR after this one is merged

**Comment by nghiant2710** [comment]
hey @thgreasi @vipulgupta2048, can you please check this PR again?

File: templates/_base-images-ref.html
Comment (thgreasi) [inline templates/_base-images-ref.html:27]: How about v ?
```suggestion
          if (IGNORE_THESE_DEVICES.includes(device.slug)) {
            return;
          }

          var cpuArch = device.is_of__cpu_architecture[0].slug;
          if (cpuArch === "rpi") {
            device.is_of__cpu_architecture[0].slug = cpuArch = "armv6hf";
          }

          if (!deviceTypesByArch[cpuArch]) {
            deviceTypesByArch[cpuArch] = [];
          }
          deviceTypesByArch[cpuArch].push(device);
```

File: templates/_base-images-ref.html
Comment (thgreasi) [reply templates/_base-images-ref.html:84]: Sorry for blocking this PR that long.
Yes, this can be a separate PR :+1: 
Let me open an issue so that we don't forget about this.

**Review by thgreasi — APPROVED** [review]
:+1:

**Comment by vipulgupta2048** [comment]
@nghiant2710 for the changes suggested above^^

**Comment by vipulgupta2048** [comment]
@balena-ci rebase

**Comment by vipulgupta2048** [comment]
@balena-ci rebase

**Comment by nghiant2710** [comment]
@vipulgupta2048, I rebased it again, should be good now.

**Pull request merged**
